### PR TITLE
Shorten the version description to the length of the DB column

### LINKED ISSUE #4188: Creating a new version may result in an exception
### Affected version(s)

4.9.26

### Description

The automatic creation of a new version may result in an exception. This is the case, if the DCA has an attribute `title`, `name`, `firstname`, `headline`, `selector` or `subject` which will be used as description for the version and this field allows more than 255 characters.

This is the part, where the attribute is read: https://github.com/contao/contao/blob/5.x/core-bundle/src/Resources/contao/classes/Versions.php#L187-L223

The description field only allows 255 chars: https://github.com/contao/contao/blob/5.x/core-bundle/src/Resources/contao/dca/tl_version.php#L60

I would suggest to shorten the description value before inserting it into the database.

I can create a PR if my suggestion is okay.

### PR DESCRIPTION
**Pull request by cliffparnitzky**
closes #4188

### CODE REVIEW

**Review by aschempp — COMMENTED** [review]
why 260 characters if the DB field is 255? Also, can we somehow dynamically determine the length if someone did increase it through a DCA change? maybe @ausi has an idea?

**Comment by ausi** [comment]
Something like this?  
`System::getContainer()->get('database_connection')->createSchemaManager()->listTableColumns('tl_version')['description']->getLength() ?? 255`

**Comment by cliffparnitzky** [comment]
> why 260 characters if the DB field is 255?
I've tested all possible values around this "255" ... and what should I say: "260" was the maximum possible.

**Comment by cliffparnitzky** [comment]
I will try the suggestion of @ausi and give feedback here.

**Comment by ausi** [comment]
> I've tested all possible values around this "255" ... and what should I say: "260" was the maximum possible.

This is probably because `StringUtil::substr()` does not separate words.

**Comment by cliffparnitzky** [comment]
Just for documentation. It has to be:
`System::getContainer()->get('database_connection')->getSchemaManager()->listTableColumns('tl_version')['description']->getLength()`.

Fixed it. Works fine.

File: core-bundle/src/Resources/contao/classes/Versions.php [hunk 1]
```diff
@@ -216,6 +216,7 @@ public function create($blnHideUser=false)
 		{
 			$strDescription = $data['subject'];
 		}
+		$strDescription = StringUtil::substr($strDescription, System::getContainer()->get('database_connection')->getSchemaManager()->listTableColumns('tl_version')['description']->getLength());
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/Versions.php]: Why not read the length from the DCA? Would save us the DB query.

Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: :-/ any suggestion?

Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: @leofeyer How should I read the length from the DCA, without String parsing the `sql` field? Or ist that the idea, you had? We cannot rely on the `maxlength` value of the `eval` array. I have no clue how to get the length value directly than via the code @ausi proposed.

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/Resources/contao/classes/Versions.php (hunk 1 above)
Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: @leofeyer Any proposal to get the conversation resolved?

**Review by aschempp — DISMISSED** [review]

File: core-bundle/src/Resources/contao/classes/Versions.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/Versions.php]: > String parsing the `sql` field? Or ist that the idea, you had?

Yes.

Comment (Toflar) [reply core-bundle/src/Resources/contao/classes/Versions.php]: We have a schema listener that does this and extracts the information to add it to the schema. It's okay to access it over the schema even if that means queries.

Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/Versions.php]:
```suggestion

		$strDescription = StringUtil::substr($strDescription, System::getContainer()->get('database_connection')->getSchemaManager()->listTableColumns('tl_version')['description']->getLength());
```

**Review by leofeyer — CHANGES_REQUESTED** [review]
I couldn‘t fix this myself because you did not allow edits by maintainers.

File: core-bundle/src/Resources/contao/classes/Versions.php [hunk 2]
```diff
@@ -217,6 +217,8 @@ public function create($blnHideUser=false)
 			$strDescription = $data['subject'];
 		}
 
+		$strDescription = StringUtil::substr($strDescription, System::getContainer()->get('database_connection')->getSchemaManager()->listTableColumns('tl_version')['description']->getLength());
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/classes/Versions.php]: Judging from the doc block `StringUtil::substr()` strips tags and may return something slightly longer than the given `$intNumberOfChars`. Is this the intended outcome here?

Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: > The function preserves words, so the result might be a bit shorter or longer than the number of characters given. It strips all tags.

I think, the length must not be longer than the detected length of the DB field.

@m-vo Any suggestion to get the ticket finally done?

Comment (ausi) [reply core-bundle/src/Resources/contao/classes/Versions.php]: > and may return something slightly longer than the given `$intNumberOfChars`.

If this is the case we should fix `StringUtil::substr()` IMO to never return something longer than `$intNumberOfChars`.

Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: Thank you @ausi . That was my thought too.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/Versions.php]: If you don‘t need the special functionality of the `StringUtil::substr()` method, please just use `mb_substr()` directly.

Comment (cliffparnitzky) [reply core-bundle/src/Resources/contao/classes/Versions.php]: Seems the version title is currently not displayed anywhere. So, the special functionality of the `StringUtil::substr()` method is not needed. I changed to code to `mb_substr()`.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @cliffparnitzky.

**Pull request merged**